### PR TITLE
Feat: handle invalid parquet data source configs

### PR DESCRIPTION
**Pull request by Meghajit**
PR for https://github.com/odpf/dagger/issues/150

### CODE REVIEW

File: dagger-core/src/test/java/io/odpf/dagger/core/source/config/adapter/SourceParquetFilePathsAdapterTest.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    @Test
+    public void shouldDeserializeJsonArrayToStringArray() throws IOException {
+        String parquetFilePathJSONString = "[\"gs://something\", \"gs://anything\"]";
+        JsonReader reader = new JsonReader(new StringReader(parquetFilePathJSONString));
+
+        SourceParquetFilePathsAdapter adapter = new SourceParquetFilePathsAdapter();
+
+        assertArrayEquals(new String[]{"gs://something", "gs://anything"}, adapter.read(reader));
+    }
+
```
Comment (prakharmathur82) [inline dagger-core/src/test/java/io/odpf/dagger/core/source/config/adapter/SourceParquetFilePathsAdapterTest.java:21]: can we add tests for empty and null inputs?

**Review by prakharmathur82 — CHANGES_REQUESTED** [review]

File: dagger-core/src/main/java/io/odpf/dagger/core/source/config/StreamConfigValidator.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.stream.Stream;
+
+import static io.odpf.dagger.core.utils.Constants.STREAM_SOURCE_DETAILS_KEY;
+import static io.odpf.dagger.core.utils.Constants.STREAM_SOURCE_PARQUET_FILE_PATHS_KEY;
+
+public class StreamConfigValidator {
+    public static StreamConfig validateSourceDetails(StreamConfig streamConfig) {
+        SourceDetails[] sourceDetailsArray = streamConfig.getSourceDetails();
+        for (SourceDetails sourceDetails : sourceDetailsArray) {
+            Preconditions.checkArgument(sourceDetails != null, "One or more elements inside %s "
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/config/StreamConfigValidator.java:17]: I didn't understand what you meant. If `SOURCE_DETAILS` config is not set or set to null, we use `SourceDetails[]{new SourceDetails(SourceName.KAFKA_CONSUMER, SourceType.UNBOUNDED)}` as the corresponding value.

This precondition is to check that each json element inside the source details array should not be null. That is, somebody should not explicitly set the config to something like this `[null]`.

File: dagger-core/src/main/java/io/odpf/dagger/core/source/config/models/TimeRanges.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package io.odpf.dagger.core.source.config.models;
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/config/models/TimeRanges.java:1]: let me check this. should be an error during of git merge conflict resolution

File: dagger-core/src/main/java/io/odpf/dagger/core/source/config/adapter/SourceParquetFilePathsAdapter.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+public class SourceParquetFilePathsAdapter extends TypeAdapter<String[]> {
+    @Override
+    public void write(JsonWriter jsonWriter, String[] strings) {
+    }
+
+    @Override
+    public String[] read(JsonReader jsonReader) throws IOException {
+        Gson gson = new Gson();
+        String[] stringArray = gson.fromJson(jsonReader, String[].class);
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/config/adapter/SourceParquetFilePathsAdapter.java:19]: Sure. Suggestions ?

File: dagger-core/src/test/java/io/odpf/dagger/core/source/config/StreamConfigTest.java [hunk 5]
```diff
@@ -256,4 +276,84 @@ public void shouldReturnEmptyTimeRangeIfParquetFileDateRangeNotGiven() {
...
+    public void shouldThrowRuntimeExceptionIfSourceDetailsArrayContainsInvalidData() {
+        when(configuration.getString(INPUT_STREAMS, ""))
+                .thenReturn("[{\"SOURCE_PARQUET_FILE_PATHS\": [\"gs://some-parquet-path\", \"gs://another-parquet-path\"],"
+                        + "\"SOURCE_PARQUET_BILLING_PROJECT\": \"data-project\","
+                        + "\"SOURCE_PARQUET_READ_ORDER_STRATEGY\": \"EARLIEST_TIME_URL_FIRST\","
+                        + "\"SOURCE_PARQUET_SCHEMA_MATCH_STRATEGY\": \"BACKWARD_COMPATIBLE_SCHEMA_WITH_FAIL_ON_TYPE_MISMATCH\","
+                        + "\"SOURCE_DETAILS\": "
+                        + "[null, {\"SOURCE_TYPE\": \"BOUNDED\", \"SOURCE_NAME\": \"PARQUET_SOURCE\"}]"
+                        + "}]");
+        assertThrows(IllegalArgumentException.class, () -> StreamConfig.parse(configuration));
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/test/java/io/odpf/dagger/core/source/config/StreamConfigTest.java:290]: I used to do this earlier actually. However, I have encountered situations earlier where I had to change the exception message for better logging requirements and then I would need to fix the breaking tests, even though just some new words got added or removed without change in logic.
Also found this, https://softwareengineerhing.stackexchange.com/questions/406079/testing-the-wording-of-an-exception-message 
Seems like there are mixed views on this.
Let me know what you think. I don't have any particular preference although.

File: dagger-core/src/test/java/io/odpf/dagger/core/source/config/adapter/SourceParquetFilePathsAdapterTest.java (hunk 1 above)
Comment (Meghajit) [reply dagger-core/src/test/java/io/odpf/dagger/core/source/config/adapter/SourceParquetFilePathsAdapterTest.java:21]: This adapter itself won't be invoked if the config is set to null. Hence, this case is checked by the `StreamConfigValidator` [here](https://github.com/odpf/dagger/pull/152/files/4ab0f5d212e375f422347dde995d50e88548accc#diff-5283f6638c0ff1eacaf3aa667f5ac3e696deec531d30c678306b939cea55d0dfR42) and is being tested [here](https://github.com/odpf/dagger/pull/152/files/4ab0f5d212e375f422347dde995d50e88548accc#diff-393cfe26f1de6214a35dcafd40df596d69b21c2fc3319a7bf86b5b051d6d5a2fR333)


For empty do you mean, what should happen if the config is be set to `[]`  ? If yes, then there is no special handling for it in the code. It will get converted to an empty string array.  I am not sure if I can write a failing test for it.

File: dagger-core/src/test/java/io/odpf/dagger/core/source/config/StreamConfigTest.java [hunk 6]
```diff
@@ -256,4 +276,84 @@ public void shouldReturnEmptyTimeRangeIfParquetFileDateRangeNotGiven() {
...
+    }
+
+    @Test
+    public void shouldThrowRuntimeExceptionIfSourceDetailsArrayContainsInvalidSourceName() {
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/test/java/io/odpf/dagger/core/source/config/StreamConfigTest.java:294]: null is an invalid source name and source type. Hence the same is verified by the tests below

File: dagger-core/src/main/java/io/odpf/dagger/core/source/config/StreamConfigValidator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (prakharmathur82) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/config/StreamConfigValidator.java:17]: it is for the error message

File: dagger-core/src/main/java/io/odpf/dagger/core/source/config/adapter/SourceParquetFilePathsAdapter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (prakharmathur82) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/config/adapter/SourceParquetFilePathsAdapter.java:19]: filePathArray

File: dagger-core/src/test/java/io/odpf/dagger/core/source/config/StreamConfigTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (prakharmathur82) [reply dagger-core/src/test/java/io/odpf/dagger/core/source/config/StreamConfigTest.java:290]: this exception can bubble up from any other place as well, so checking for message can help

File: dagger-core/src/main/java/io/odpf/dagger/core/source/config/models/TimeRanges.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/config/models/TimeRanges.java]: Fixed via commit https://github.com/odpf/dagger/pull/152/commits#:~:text=feat%3A%20delete%20erroneously%20added%20class%20and%20test

File: dagger-core/src/test/java/io/odpf/dagger/core/source/config/models/TimeRangesTest.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
+package io.odpf.dagger.core.source.config.models;
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/test/java/io/odpf/dagger/core/source/config/models/TimeRangesTest.java]: Fixed via commit https://github.com/odpf/dagger/pull/152/commits/bd5ba02b0b01606dfccc6e3cd88436f67ff12628

File: dagger-core/src/test/java/io/odpf/dagger/core/source/config/adapter/SourceParquetFilePathsAdapterTest.java (hunk 1 above)
Comment (Meghajit) [reply dagger-core/src/test/java/io/odpf/dagger/core/source/config/adapter/SourceParquetFilePathsAdapterTest.java:21]: Fixed via commit https://github.com/odpf/dagger/pull/152/commits/c8972fcebe55d361762ab259fd9f462a79b68c74

File: dagger-core/src/main/java/io/odpf/dagger/core/source/config/adapter/SourceParquetFilePathsAdapter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/config/adapter/SourceParquetFilePathsAdapter.java]: Fixed via commit https://github.com/odpf/dagger/pull/152/commits/c8972fcebe55d361762ab259fd9f462a79b68c74

File: dagger-core/src/test/java/io/odpf/dagger/core/source/config/StreamConfigTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/test/java/io/odpf/dagger/core/source/config/StreamConfigTest.java]: Fixed via commit https://github.com/odpf/dagger/pull/152/commits/9088f63e178e1dae7d63d8db11cc16fa214c5c59

File: dagger-core/src/test/java/io/odpf/dagger/core/source/config/StreamConfigTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/test/java/io/odpf/dagger/core/source/config/StreamConfigTest.java:310]: Fixed via commit https://github.com/odpf/dagger/pull/152/commits/9088f63e178e1dae7d63d8db11cc16fa214c5c59

**Review by prakharmathur82 — APPROVED** [review]
